Fix #678 : Landscape onBoarding workflow

Sarthak2601 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
- Fixes part of #678 
- Added the svg files for the landscape view
- One of the svg file contained an image link so replaced that with path source and everything to make that image come back
- Handled the source of the slide image using the orientation configuration of android in OnboadingSlideViewModel.kt
- Changed the ScreenOrientation of the onBoarding activity in the Manifest

## Screenshots - 
<img src="https://user-images.githubusercontent.com/53938155/75097656-473f1b80-55d3-11ea-9072-237d98c11bcc.png" width=200/>
<img src="https://user-images.githubusercontent.com/53938155/75097658-4b6b3900-55d3-11ea-821c-d3d184b08567.png" width=200/>

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Sarthak2601 committed:
Landscape onBoarding XML

Sarthak2601 commented:
@rt4914 PTAL :-)

rt4914 commented on app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt:
```diff
@@ -14,6 +16,7 @@ class OnboardingSlideViewModel(val context: Context, viewPagerSlide: ViewPagerSl
   val contentDescription = ObservableField<String>(context.resources.getString(R.string.onboarding_slide_0_title))
   val title = ObservableField<String>(context.resources.getString(R.string.onboarding_slide_0_title))
   val description = ObservableField<String>(context.resources.getString(R.string.onboarding_slide_0_description))
+  val orientation = Resources.getSystem().configuration.orientation
```
Make this private

Sarthak2601 replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt:
```diff
@@ -22,25 +25,29 @@ class OnboardingSlideViewModel(val context: Context, viewPagerSlide: ViewPagerSl
       ViewPagerSlide.SLIDE_0 -> {
-        slideImage.set(R.drawable.ic_onboarding_0)
+        if(orientation == Configuration.ORIENTATION_LANDSCAPE){slideImage.set(R.drawable.ic_landscape_onboarding_2)}
+        else if(orientation == Configuration.ORIENTATION_PORTRAIT){slideImage.set(R.drawable.ic_onboarding_0)}
```
Write this if else code in this format:
```
       if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
          slideImage.set(R.drawable.ic_landscape_onboarding_2)
        } else if (orientation == Configuration.ORIENTATION_PORTRAIT) {
          slideImage.set(R.drawable.ic_onboarding_0)
        }
```

rt4914 replied:
@Sarthak2601 Once you have solved this, please reply to this comment. In this case you can simply write `Done`, ditto for other comments too.

Sarthak2601 replied:
Okay 👍
Done :-)

rt4914 commented on app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt:
```diff
@@ -22,25 +25,29 @@ class OnboardingSlideViewModel(val context: Context, viewPagerSlide: ViewPagerSl
     when (viewPagerSlide) {
       ViewPagerSlide.SLIDE_0 -> {
-        slideImage.set(R.drawable.ic_onboarding_0)
+        if(orientation == Configuration.ORIENTATION_LANDSCAPE){slideImage.set(R.drawable.ic_landscape_onboarding_2)}
```
The final/image numbering is incorrect. The first slide should have 0 index, second slide should have 1 index and so on, similar to previous images. Here it should be `ic_onboarding_landscape_0`

Sarthak2601 replied:
Done

rt4914 commented on app/src/main/res/layout-land/onboarding_fragment.xml:
```diff
@@ -0,0 +1,97 @@
+      app:layout_constraintStart_toStartOf="parent"
+      app:layout_constraintTop_toTopOf="parent">
+
+
```
Remove this extra blank lines.

Sarthak2601 replied:
Done 👍

rt4914 commented on app/src/main/res/layout-land/onboarding_fragment.xml:
```diff
@@ -0,0 +1,97 @@
+      app:layout_constraintBottom_toTopOf="@id/bottom_container"
+      app:layout_constraintEnd_toEndOf="parent"
+      app:layout_constraintStart_toStartOf="parent"
+      app:layout_constraintTop_toTopOf="parent">
```
We can make this a self-closing tag 
`app:layout_constraintTop_toTopOf="parent" />`

Sarthak2601 replied:
Done 👍

rt4914 reviewed (changes requested):
The initial structure looks correct but if you look at the mock correctly, the left side of screen contains the image background.

Mock:
<img width="641" alt="Screenshot 2020-02-23 at 11 45 22 AM" src="https://user-images.githubusercontent.com/9396084/75104689-19d58a80-5632-11ea-8bcc-fef46ff4945b.png">

App:
<img width="585" alt="Screenshot 2020-02-23 at 11 45 09 AM" src="https://user-images.githubusercontent.com/9396084/75104688-17733080-5632-11ea-993e-3d13f0b91a22.png">

Sarthak2601 commented:
> The initial structure looks correct but if you look at the mock correctly, the left side of screen contains the image background.
> 
> Mock:
> <img alt="Screenshot 2020-02-23 at 11 45 22 AM" width="641" src="https://user-images.githubusercontent.com/9396084/75104689-19d58a80-5632-11ea-8bcc-fef46ff4945b.png">
> 
> App:
> <img alt="Screenshot 2020-02-23 at 11 45 09 AM" width="585" src="https://user-images.githubusercontent.com/9396084/75104688-17733080-5632-11ea-993e-3d13f0b91a22.png">

Actually, I wanted to discuss this with you. The thing here is that we have the constraint layout with the skip button and the dots at the bottom in the fragment and the slide gets loaded in the view pager above that. As the slides load in the view pager it can't be continued till the end because of the presence of the another constraint layout.

PrarabdhGarg commented:
> > The initial structure looks correct but if you look at the mock correctly, the left side of screen contains the image background.
> > Mock:
> > <img alt="Screenshot 2020-02-23 at 11 45 22 AM" width="641" src="https://user-images.githubusercontent.com/9396084/75104689-19d58a80-5632-11ea-8bcc-fef46ff4945b.png">
> > App:
> > <img alt="Screenshot 2020-02-23 at 11 45 09 AM" width="585" src="https://user-images.githubusercontent.com/9396084/75104688-17733080-5632-11ea-993e-3d13f0b91a22.png">
> 
> Actually, I wanted to discuss this with you. The thing here is that we have the constraint layout with the skip button and the dots at the bottom in the fragment and the slide gets loaded in the view pager above that. As the slides load in the view pager it can't be continued till the end because of the presence of the another constraint layout.

@Sarthak2601 what you could do for that is that instead of constraining the bottom of the viewpager with the top of the bottom container, you could constraint it with the bottle of the parent. That would give it the necessary overlap required.

Sarthak2601 committed:
Resolving issues

Sarthak2601 committed:
Resolving issues

Sarthak2601 commented:
> > > The initial structure looks correct but if you look at the mock correctly, the left side of screen contains the image background.
> > > Mock:
> > > <img alt="Screenshot 2020-02-23 at 11 45 22 AM" width="641" src="https://user-images.githubusercontent.com/9396084/75104689-19d58a80-5632-11ea-8bcc-fef46ff4945b.png">
> > > App:
> > > <img alt="Screenshot 2020-02-23 at 11 45 09 AM" width="585" src="https://user-images.githubusercontent.com/9396084/75104688-17733080-5632-11ea-993e-3d13f0b91a22.png">
> > 
> > 
> > Actually, I wanted to discuss this with you. The thing here is that we have the constraint layout with the skip button and the dots at the bottom in the fragment and the slide gets loaded in the view pager above that. As the slides load in the view pager it can't be continued till the end because of the presence of the another constraint layout.
> 
> @Sarthak2601 what you could do for that is that instead of constraining the bottom of the viewpager with the top of the bottom container, you could constraint it with the bottle of the parent. That would give it the necessary overlap required.

Oh yes, found that. Thank you so much @PrarabdhGarg :-)

Sarthak2601 committed:
get started button positioning

Sarthak2601 commented:
Updated the positioning of the get started button in accordance with the mock. 
<img src="https://user-images.githubusercontent.com/53938155/75105648-f0b9f780-563b-11ea-9df1-24bd78a5a144.png" width=200/>

@rt4914 Kindly review.

rt4914 reviewed (changes requested):
@Sarthak2601 Looks good, thanks. 
Update the title of this PR: `Fix #678: Landscape onBoarding workflow`

Also, now we can focus on test cases.

Check `OnboardingFragmentTest` in this you will find a lot of test cases for portrait mode. In this file you will need add following test cases for orientation mode.

1. Rotate to orientation mode, check slide_0 title, if this is correct or not.
2. In portrait mode, move to slide_1 , change to landscape mode and now check if the title is correct or not.
3. In portrait mode, click on skip, change to landscape mode and now check the title is same as last slide title or not.


Regarding how to change to orientation landscape in a test case, you can see test cases in `AdminAuthActivityTest`

Sarthak2601 commented:
> @Sarthak2601 Looks good, thanks.
> Update the title of this PR: `Fix #678: Landscape onBoarding workflow`
> 
> Also, now we can focus on test cases.
> 
> Check `OnboardingFragmentTest` in this you will find a lot of test cases for portrait mode. In this file you will need add following test cases for orientation mode.
> 
> 1. Rotate to orientation mode, check slide_0 title, if this is correct or not.
> 2. In portrait mode, move to slide_1 , change to landscape mode and now check if the title is correct or not.
> 3. In portrait mode, click on skip, change to landscape mode and now check the title is same as last slide title or not.
> 
> Regarding how to change to orientation landscape in a test case, you can see test cases in `AdminAuthActivityTest`

Okay,  I will work on it 👍
Thank you so much for the guidance :-)

Sarthak2601 commented:
> The initial structure looks correct but if you look at the mock correctly, the left side of screen contains the image background.
> 
> Mock:
> <img alt="Screenshot 2020-02-23 at 11 45 22 AM" width="641" src="https://user-images.githubusercontent.com/9396084/75104689-19d58a80-5632-11ea-8bcc-fef46ff4945b.png">
> 
> App:
> <img alt="Screenshot 2020-02-23 at 11 45 09 AM" width="585" src="https://user-images.githubusercontent.com/9396084/75104688-17733080-5632-11ea-993e-3d13f0b91a22.png">

Done 👍

Sarthak2601 committed:
Adding test cases

Sarthak2601 commented:
> @Sarthak2601 Looks good, thanks.
> Update the title of this PR: `Fix #678: Landscape onBoarding workflow`
> 
> Also, now we can focus on test cases.
> 
> Check `OnboardingFragmentTest` in this you will find a lot of test cases for portrait mode. In this file you will need add following test cases for orientation mode.
> 
> 1. Rotate to orientation mode, check slide_0 title, if this is correct or not.
> 2. In portrait mode, move to slide_1 , change to landscape mode and now check if the title is correct or not.
> 3. In portrait mode, click on skip, change to landscape mode and now check the title is same as last slide title or not.
> 
> Regarding how to change to orientation landscape in a test case, you can see test cases in `AdminAuthActivityTest`

-Added test cases 
-Changed the PR title 

@rt4914, PTAL :-)

rt4914 commented on app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt:
```diff
@@ -437,4 +439,39 @@ class OnboardingFragmentTest {
+      onView(withId(R.id.slide_title_text_view))
+        .check(
+          matches(
+            withText(R.string.onboarding_slide_3_title)))
```
Indentation change suggested:
```
      onView(withId(R.id.skip_text_view)).perform(click())
      onView(isRoot()).perform(orientationLandscape())
      onView(withId(R.id.slide_title_text_view)).check(matches(withText(R.string.onboarding_slide_3_title)))
```

rt4914 replied:
Ditto for above two cases also.

Sarthak2601 replied:
> These 3 test cases are failing.
> Update them as following:
> 
> ```
>   @Test
>   fun testOnboardingFragment_slide0title_inlandscape_iscorrectornot() {
>     launch(OnboardingActivity::class.java).use {
>       onView(isRoot()).perform(orientationLandscape())
>       onView(
>         allOf(
>           withId(R.id.slide_title_text_view),
>           isCompletelyDisplayed()
>         )
>       ).check(matches(withText(R.string.onboarding_slide_0_title)))
>     }
>   }
> 
>   @Test
>   fun testOnboardingFragment_movetoslide1_changeorientation_checktitle() {
>     launch(OnboardingActivity::class.java).use {
>       onView(withId(R.id.onboarding_slide_view_pager)).perform(scrollToPage(1))
>       onView(isRoot()).perform(orientationLandscape())
>       onView(
>         allOf(
>           withId(R.id.slide_title_text_view),
>           isCompletelyDisplayed()
>         )
>       ).check(matches(withText(R.string.onboarding_slide_1_title)))
>     }
>   }
> 
>   @Test
>   fun testOnboardingFragment_clickonskip_changeorientation_checktitle() {
>     launch(OnboardingActivity::class.java).use {
>       onView(withId(R.id.skip_text_view)).perform(click())
>       onView(isRoot()).perform(orientationLandscape())
>       onView(
>         allOf(
>           withId(R.id.slide_title_text_view),
>           isCompletelyDisplayed()
>         )
>       ).check(matches(withText(R.string.onboarding_slide_3_title)))
>     }
>   }
> ```

Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt:
```diff
@@ -437,4 +439,39 @@ class OnboardingFragmentTest {
+  }
+
+  @Test
+  fun testOnboardingFragment_clickonskip_changeorientation_checktitle() {
```
Test case name:
`testOnboardingFragment_clickOnSkip_changeOrientation_titleIsCorrect`

Sarthak2601 replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt:
```diff
@@ -437,4 +439,39 @@ class OnboardingFragmentTest {
+  }
+
+  @Test
+  fun testOnboardingFragment_movetoslide1_changeorientation_checktitle() {
```
Test case name:
`testOnboardingFragment_moveToSlide1_changeOrientation_titleIsCorrect`

Sarthak2601 replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt:
```diff
@@ -437,4 +439,39 @@ class OnboardingFragmentTest {
   }
+
+  @Test
+  fun testOnboardingFragment_slide0title_inlandscape_iscorrectornot(){
```
Nit:
`testOnboardingFragment_slide0Title_changeOrientation_titleIsCorrect `

Sarthak2601 replied:
Done.

rt4914 reviewed (changes requested):
Some small changes suggested.

rt4914 reviewed (changes requested):
These 3 test cases are failing.
Update them as following:
```
  @Test
  fun testOnboardingFragment_slide0title_inlandscape_iscorrectornot() {
    launch(OnboardingActivity::class.java).use {
      onView(isRoot()).perform(orientationLandscape())
      onView(
        allOf(
          withId(R.id.slide_title_text_view),
          isCompletelyDisplayed()
        )
      ).check(matches(withText(R.string.onboarding_slide_0_title)))
    }
  }

  @Test
  fun testOnboardingFragment_movetoslide1_changeorientation_checktitle() {
    launch(OnboardingActivity::class.java).use {
      onView(withId(R.id.onboarding_slide_view_pager)).perform(scrollToPage(1))
      onView(isRoot()).perform(orientationLandscape())
      onView(
        allOf(
          withId(R.id.slide_title_text_view),
          isCompletelyDisplayed()
        )
      ).check(matches(withText(R.string.onboarding_slide_1_title)))
    }
  }

  @Test
  fun testOnboardingFragment_clickonskip_changeorientation_checktitle() {
    launch(OnboardingActivity::class.java).use {
      onView(withId(R.id.skip_text_view)).perform(click())
      onView(isRoot()).perform(orientationLandscape())
      onView(
        allOf(
          withId(R.id.slide_title_text_view),
          isCompletelyDisplayed()
        )
      ).check(matches(withText(R.string.onboarding_slide_3_title)))
    }
  }
```

Sarthak2601 committed:
Resolving issues

rt4914 reviewed (approved):
LGTM. Thanks.

Sarthak2601 commented:
> LGTM. Thanks.

Thank you so much ! 
You helped me a lot in this :-)

rt4914 commented:
> > LGTM. Thanks.
> 
> Thank you so much !
> You helped me a lot in this :-)

No worries. I will mostly assign you some new tasks to work on tomorrow.

Sarthak2601 commented:
> > > LGTM. Thanks.
> > 
> > 
> > Thank you so much !
> > You helped me a lot in this :-)
> 
> No worries. I will mostly assign you some new tasks to work on tomorrow.

Sounds great !
